airflow/www/views.py

```diff
@@ -678,18 +678,24 @@ def index(self):
                 for name, in dagtags
             ]
 
-            import_errors = session.query(errors.ImportError).order_by(errors.ImportError.id).all()
+            if (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG) in user_permissions:
+                import_errors = session.query(errors.ImportError).order_by(errors.ImportError.id).all()
 
-        if import_errors:
-            dag_filenames = {dag.fileloc for dag in dags}
-            all_dags_readable = (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG) in user_permissions
+            else:
+                # if the user doesn't have access to all DAGs, only display errors from visible DAGs
+                import_errors = (
+                    session.query(errors.ImportError)
+                    .join(DagModel, DagModel.fileloc == errors.ImportError.filename)
+                    .filter(DagModel.dag_id.in_(filter_dag_ids))
+                    .order_by(errors.ImportError.id)
+                ).all()
 
+        if import_errors:
             for import_error in import_errors:
```

Comment:
```suggestion
        for import_error in import_errors:
```

We can get rid of this `if` now.